# Stage 3: Exploratory Data Analysis (EDA)

## Objective

The purpose of Exploratory Data Analysis (EDA) is to discover patterns,
relationships and factors associated with employee attrition.

The analysis will focus on identifying the characteristics of employees
who are more likely to leave the company.

In [2]:
# Import required libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:
# Load cleaned dataset

df = pd.read_csv("../data/CleanedDataset/employee_cleaned.csv")

# Attrition Rate Function

To avoid repeating calculations, a reusable function is created.

The function calculates attrition percentages for any categorical column.

In [4]:
# Function to calculate attrition rates

def attrition_rate(df, column):

    result = (
        pd.crosstab(
            df[column],
            df["Attrition"],
            normalize="index"
        ) * 100
    )

    return result.round(2)

In [5]:
#test call to the function
attrition_rate(df, "Department")

Attrition,No,Yes
Department,,
Human Resources,80.95,19.05
Research & Development,86.16,13.84
Sales,79.37,20.63


### Observation

The function returns attrition percentages
for every category within a selected feature.

This function will be reused throughout the project.

# Attrition Distribution

### Business Question

What proportion of employees left the company?

In [6]:
df["Attrition"].value_counts()

Attrition
No     1233
Yes     237
Name: count, dtype: int64

In [7]:
round(
    df["Attrition"].value_counts(normalize=True) * 100,
    2
)
#passing normalize parameter to get a result in proportion (%), if normalize false means than it will return the count

Attrition
No     83.88
Yes    16.12
Name: proportion, dtype: float64

### Observation

Most employees remained in the company.

A smaller percentage of employees left the company.

### Business Insight

Employee attrition exists within the organization and should be
investigated to identify contributing factors.

# Department vs Attrition

### Business Question

Which department experiences the highest employee attrition?

In [8]:
attrition_rate(df, "Department")

Attrition,No,Yes
Department,,
Human Resources,80.95,19.05
Research & Development,86.16,13.84
Sales,79.37,20.63


### Observation

Sales department shows the highest attrition rate.

Research & Development shows the lowest attrition rate among
major departments.

### Business Insight

Retention strategies should be prioritized within the Sales department.

# Overtime vs Attrition

### Business Question

Do employees working overtime leave more often?

In [9]:
attrition_rate(df, "OverTime")

Attrition,No,Yes
OverTime,,
No,89.56,10.44
Yes,69.47,30.53


### Observation

Employees working overtime have a substantially higher attrition rate.

### Business Insight

Overtime appears to be one of the strongest factors associated
with employee attrition.

# Job Satisfaction vs Attrition

In [10]:
attrition_rate(df, "JobSatisfaction")

Attrition,No,Yes
JobSatisfaction,,
1,77.16,22.84
2,83.57,16.43
3,83.48,16.52
4,88.67,11.33


### Observation

Employees with less job satisfaction have a substantially higher attrition rate.

### Business Insight
Less Job satisfaction to be one of the strongest factors associated with employee attrition.

# Work-Life Balance vs Attrition

In [11]:
attrition_rate(df, "WorkLifeBalance")

Attrition,No,Yes
WorkLifeBalance,,
1,68.75,31.25
2,83.14,16.86
3,85.78,14.22
4,82.35,17.65


### Observation

Employees with less work life balance have a substantially higher attrition rate.

### Business Insight
Work life balance to be one of the strongest factors associated with employee attrition.

# Business Travel vs Attrition

In [12]:
attrition_rate(df, "BusinessTravel")

Attrition,No,Yes
BusinessTravel,,
Non-Travel,92.00,8.00
Travel_Frequently,75.09,24.91
Travel_Rarely,85.04,14.96


### Observation

Employees who travel frequently have a substantially higher attrition rate.

### Business Insight
Employees travel frequently to be one of the strongest factors associated with employee attrition.

# Job Role vs Attrition

In [13]:
attrition_rate(df, "JobRole")

Attrition,No,Yes
JobRole,,
Healthcare Representative,93.13,6.87
Human Resources,76.92,23.08
Laboratory Technician,76.06,23.94
Manager,95.10,4.90
Manufacturing Director,93.10,6.90
Research Director,97.50,2.50
Research Scientist,83.90,16.10
Sales Executive,82.52,17.48
Sales Representative,60.24,39.76


### Observation

Sales Representative have a substantially higher attrition rate.
Research Director have a very less attrition rate

### Business Insight
Employees working as Sales Representative to be one of the strongest factors associated with employee attrition.

# Monthly Income vs Attrition

### Business Question

Do lower-paid employees leave more frequently?

In [14]:
df.groupby(
    "Attrition"
)["MonthlyIncome"].mean()

Attrition
No     6832.739659
Yes    4787.092827
Name: MonthlyIncome, dtype: float64

### Observation

Employees who left have a lower average monthly income.

### Business Insight

Compensation may influence employee retention.

# Experience Analysis

In [15]:
df.groupby(
    "Attrition"
)["TotalWorkingYears"].mean().round(2)

Attrition
No     11.86
Yes     8.24
Name: TotalWorkingYears, dtype: float64

### Observation
Employees who stayed have
an average of 11.86 years
of total work experience.

Employees who left have
an average of 8.24 years
of total work experience.

# Business Insight
Less experienced employees appear
more likely to leave the company.

This suggests that early-career
employees may require additional
retention efforts.

# Company Tenure Analysis (Years at Company vs Attrition)

In [16]:
df.groupby(
    "Attrition"
)["YearsAtCompany"].mean().round(1)

Attrition
No     7.4
Yes    5.1
Name: YearsAtCompany, dtype: float64

### Observation

Employees who left generally spent fewer years at the company.

# Save Important Results

In [17]:
import os

os.makedirs("results", exist_ok=True)

In [18]:
attrition_rate(df, "Department").to_csv(
    "../results/department_attrition.csv"
)


# EDA Summary & Conclusion

## Key Insights

- **Overall Attrition:** The company has an employee attrition rate of **16.12%**.
- **Highest Risk Department:** The **Sales Department** shows the highest attrition rate (**20.63%**) among all departments.
- **Highest Risk Role:** **Sales Representatives** exhibit the highest attrition rate (**39.76%**) among all job roles.
- **Burnout Factors:** Employees working **Overtime** have an attrition rate of **30.53%**, while employees who travel **Frequently** for business show an attrition rate of **24.91%**.
- **Employee Well-being:** Lower **Work-Life Balance** and **Job Satisfaction** levels are associated with higher attrition rates.
- **Compensation & Experience:** Employees who left earned an average monthly income of **₹4,787.09**, compared to **₹6,832.74** for employees who stayed.
- **Experience & Tenure:** Employees who left had an average of **8.24 years** of work experience and **5.13 years** at the company, compared to **11.86 years** and **7.37 years** respectively for retained employees.

---

## Conclusion & Recommendations

The analysis indicates that employee attrition is strongly associated with overtime work, frequent business travel, lower satisfaction levels, lower compensation, and shorter tenure. Sales-related roles, particularly Sales Representatives, exhibit the highest turnover rates.

### Recommendations

1. **Prioritize Sales Retention:** Develop targeted retention programs for Sales Representatives and employees in the Sales department.
2. **Reduce Burnout:** Monitor overtime workloads and explore alternatives to frequent business travel where possible.
3. **Improve Employee Well-being:** Strengthen work-life balance initiatives and employee satisfaction programs.
4. **Support Early-Career Employees:** Review compensation structures, onboarding processes, and career development opportunities for less experienced employees.

# Correlation Analysis

In [25]:
df["AttritionNumeric"] = df["Attrition"].map({
    "No": 0,
    "Yes": 1
})

In [26]:
numerical_df = df.select_dtypes(include="number")

attrition_corr = (
    numerical_df
    .corr()["AttritionNumeric"]
    .drop("AttritionNumeric")
    .sort_values(ascending=False)
)

print(attrition_corr)

DistanceFromHome            0.077924
NumCompaniesWorked          0.043494
MonthlyRate                 0.015170
PerformanceRating           0.002889
HourlyRate                 -0.006846
PercentSalaryHike          -0.013478
Education                  -0.031373
YearsSinceLastPromotion    -0.033019
RelationshipSatisfaction   -0.045872
DailyRate                  -0.056652
TrainingTimesLastYear      -0.059478
WorkLifeBalance            -0.063939
EnvironmentSatisfaction    -0.103369
JobSatisfaction            -0.103481
JobInvolvement             -0.130016
YearsAtCompany             -0.134392
StockOptionLevel           -0.137145
YearsWithCurrManager       -0.156199
Age                        -0.159205
MonthlyIncome              -0.159840
YearsInCurrentRole         -0.160545
JobLevel                   -0.169105
TotalWorkingYears          -0.171063
Name: AttritionNumeric, dtype: float64


### Correlation Analysis Summary

Correlation analysis revealed that employee attrition is most negatively associated with Total Working Years (-0.171), Job Level (-0.169), Years in Current Role (-0.161), Monthly Income (-0.160), and Age (-0.159).

These findings indicate that employees who are more experienced, occupy higher job positions, earn higher salaries, and have longer tenure are generally less likely to leave the organization.

Distance From Home (0.078) showed the strongest positive correlation with attrition, suggesting that employees living farther from the workplace may have a slightly higher tendency to leave.